README.md
<!-- badges: start -->

[![R build
status](https://github.com/federicomarini/HCAData/workflows/R-CMD-check/badge.svg)](https://github.com/federicomarini/HCAData/actions)
<!-- badges: end -->

HCAData
=======

Available at <http://bioconductor.org/packages/HCAData/>.

This is an
R/[ExperimentHub](https://bioconductor.org/packages/release/bioc/html/ExperimentHub.html)
package for accessing data from the [Human Cell
Atlas](https://www.humancellatlas.org) as
[`SingleCellExperiment`](https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html)
objects.

The datasets can otherwise be downloaded from the [Human Cell Atlas Data
Portal](https://preview.data.humancellatlas.org), where they are
available to the research community, either in raw format (with scripts
to download the primary data), or in hdf5-based format (for the sets
generated via the 10x method).

If you use the data, please refer to the original sources as well (plus
the related publications, which will be listed here when they will be
out), which are licensed under a Creative Commons Attribution 4.0
International License.

Some additional information
---------------------------

-   The
    [`cellranger`](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger)-processed
    data files and metadata were downloaded from the links provided in
    [the Data Portal](https://preview.data.humancellatlas.org) on
    November 5, 2018

-   This current release includes the following datasets:

| dataset.ID             | subset               | species      | technology | cell.number | description                                                                                 |
|:-----------------------|:---------------------|:-------------|:-----------|------------:|:--------------------------------------------------------------------------------------------|
| Census of Immune Cells | Umbilical cord blood | Homo sapiens | 10x        |      384000 | Profiling of immunocytes by single cell RNA-seq for understanding human health and disease. |
| Census of Immune Cells | Bone marrow          | Homo sapiens | 10x        |      378000 | Profiling of immunocytes by single cell RNA-seq for understanding human health and disease. |

-   Since the hdf5 files are derived from the cellranger pipeline, some
    information on their specific format can be found
    [here](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/advanced/h5_matrices)

Some new dataset has been released but it is not in this package!
-----------------------------------------------------------------

In this case, feel free to contact the author(s) of this package, and/or
to provide a complete pull request with the recipe to add the missing
dataset.

To get initial guidance for this, please refer to the excellent
vignettes of the
[ExperimentHub](https://bioconductor.org/packages/release/bioc/html/ExperimentHub.html)
package.


NEWS.md
# HCAData 1.8.0

## Other notes

- The functions to load the data gain a `as.sparse` parameter, to control whether the underlying HDF5 dataset should be treated as sparse or not.

# HCAData 1.0.0

## Other notes

- `HCAData` is on Bioconductor!

# HCAData 0.1.0

## New features

- Initial version of the package!


tests/testthat.R
library(testthat)
library(HCAData)

test_check("HCAData")


tests/testthat/test-object_check.R
context("sce objects are correct")

test_that("empty call", {
  library(HCAData)
  expect_message(HCAData())
})

test_that("wrong call to the dataset", {
  library(HCAData)
  expect_error(HCAData("cordblood"))
})

test_that("cord blood data is correct", {
  library(HCAData)
  sce_cordblood <- HCAData("ica_cord_blood")
  expect_is(sce_cordblood, "SingleCellExperiment")
  expect_equal(nrow(sce_cordblood), 33694)
  expect_equal(ncol(sce_cordblood), 384000)
  expect_equal(colnames(rowData(sce_cordblood)), c("ID", "Symbol"))
  expect_equal(assayNames(sce_cordblood), "counts")
  expect_equal(length(grep("^ERCC-", rownames(sce_cordblood))), 0)
})

test_that("bone marrow data is correct", {
  library(HCAData)
  sce_bonemarrow <- HCAData("ica_bone_marrow")
  expect_is(sce_bonemarrow, "SingleCellExperiment")
  expect_equal(nrow(sce_bonemarrow), 33694)
  expect_equal(ncol(sce_bonemarrow), 378000)
  expect_equal(colnames(rowData(sce_bonemarrow)), c("ID", "Symbol"))
  expect_equal(assayNames(sce_bonemarrow), "counts")
  expect_equal(length(grep("^ERCC-", rownames(sce_bonemarrow))), 0)
})


R/HCAData-pkg.R
#' HCAData: Accessing The Datasets Of The Human Cell Atlas in R/Bioconductor
#'
#' \code{HCAData} is an ExperimentHub package which provides access to the single-cell
#' RNA-seq data from the Human Cell Atlas project (\url{https://www.humancellatlas.org})
#'
#' @references If you use the data in this package, please refer to the original sources
#' (Human Cell Atlas Data Portal, \url{https://preview.data.humancellatlas.org}) as well
#' (plus the related publications, which will be listed here when they will be out),
#' which are licensed under a Creative Commons Attribution 4.0 International License.
#'
#' @importFrom utils read.csv
#' @importFrom ExperimentHub createHubAccessors ExperimentHub
#' @importFrom AnnotationHub query
#' @importFrom HDF5Array HDF5Array
#' @importFrom SingleCellExperiment SingleCellExperiment
#'
#' @author Federico Marini \email{marinif@@uni-mainz.de}
#' @name HCAData-pkg
#' @docType package
NULL


inst/scripts/make-metadata.R
## validate with `ExperimentHubData::makeExperimentHubMetadata()`
## (above pkg directory)

df_HCA_cordblood <- data.frame(
  Title = paste0("Human Cell Atlas - Census of Immune Cells, Umbilical cord blood, ",
                 c("'dense matrix' format","sample (column) annotation","gene (row) annotation")),
  Description = paste0("Census of Immune Cells, Umbilical cord blood - 10x data from the Human Cell Atlas project, profiling of immunocytes by single cell RNA-seq for understanding human health and disease. Represented as HDF5 count matrix and serialized rowData and colData objects, derived from the raw counts matrix on https://preview.data.humancellatlas.org - ",
                       c("Full rectangular, block-compressed format, 1GB block size.",
                         "Sample descriptions (barcodes).",
                         "Gene descriptions.")),
  BiocVersion = rep("3.9", 3),
  Genome = rep("GRCh38", 3),
  SourceType = rep("HDF5", 3),
  SourceUrl = rep("https://s3.amazonaws.com/preview-ica-expression-data/ica_cord_blood_h5.h5", 3),
  SourceVersion = rep("b5dd8ec9c9ff2d142986aaee014b9098", 3),
  Species = rep("Homo sapiens", 3),
  TaxonomyId = rep(9606, 3),
  Coordinate_1_based = rep(TRUE, 3),
  DataProvider = rep("Human Cell Atlas", 3),
  Maintainer = rep("Federico Marini <marinif@uni-mainz.de>", 3),
  RDataClass = rep("character", 3),
  DispatchClass = c("H5File","Rds","Rds"),
  RDataPath = paste0("HCAData/ica_cord_blood",c("_rectangular.h5", "_colData.rds", "_rowData.rds")),
  stringsAsFactors = FALSE
)

df_HCA_bonemarrow <- data.frame(
  Title = paste0("Human Cell Atlas - Census of Immune Cells, Bone marrow, ",
                 c("'dense matrix' format","sample (column) annotation","gene (row) annotation")),
  Description = paste0("Census of Immune Cells, Bone marrow - 10x data from the Human Cell Atlas project, profiling of immunocytes by single cell RNA-seq for understanding human health and disease. Represented as HDF5 count matrix and serialized rowData and colData objects, derived from the raw counts matrix on https://preview.data.humancellatlas.org - ",
                       c("Full rectangular, block-compressed format, 1GB block size.",
                         "Sample descriptions (barcodes).",
                         "Gene descriptions.")),
  BiocVersion = rep("3.9", 3),
  Genome = rep("GRCh38", 3),
  SourceType = rep("HDF5", 3),
  SourceUrl = rep("https://s3.amazonaws.com/preview-ica-expression-data/ica_bone_marrow_h5.h5", 3),
  SourceVersion = rep("9d10973b7c178ee6886ae12d004d4d23", 3),
  Species = rep("Homo sapiens", 3),
  TaxonomyId = rep(9606, 3),
  Coordinate_1_based = rep(TRUE, 3),
  DataProvider = rep("Human Cell Atlas", 3),
  Maintainer = rep("Federico Marini <marinif@uni-mainz.de>", 3),
  RDataClass = rep("character", 3),
  DispatchClass = c("H5File","Rds","Rds"),
  RDataPath = paste0("HCAData/ica_bone_marrow",c("_rectangular.h5", "_colData.rds", "_rowData.rds")),
  stringsAsFactors = FALSE
)

# when there's going to be many more, it might be practical to have it automated

## Combined meta-data
df_all <- do.call(rbind,lapply(ls(pattern = "^df_HCA"),get))

## Save .csv file
write.csv(df_all, file = "../extdata/metadata.csv", row.names = FALSE)



R/HCAData.R
#' @title Download data from the HCA via ExperimentHub
#'
#' @description Download HDF5 (dense assay) and RDS (row and column annotations)
#' files from the HCA via ExperimentHub, composing them together as a
#' SingleCellExperiment object
#'
#' @details This current release includes the following datasets:
#' \describe{
#'   \item{Census of Immune Cells - Umbilical cord blood}{UMI counts from the 10x
#'   (droplet) single-cell RNA-seq data. The object contains counts for 33694 genes
#'   in 384000 cells.}
#'   \item{Census of Immune Cells - Bone marrow}{UMI counts from the 10x
#'   (droplet) single-cell RNA-seq data. The object contains counts for 33694 genes
#'   in 378000 cells.}
#' }
#'
#' @param dataset A character string: which dataset should be retrieved?
#' @param as.sparse Logical, specifies whether the underlying HDF5 dataset should
#' be treated as sparse or not - will be passed to the call to `HDF5Array()`.
#' Defaults to TRUE, i.e. by using the `DelayedArray` infrastructure.
#'
#' @return A SingleCellExperiment object with a HDF5Matrix in the \code{counts}
#' assay, containing the UMI counts for each gene in each cell. Row- and column-level
#' metadata are also provided.
#'
#' @export
#'
#' @examples
#' HCAData()
#' sce_cordblood <- HCAData("ica_cord_blood")
HCAData <- function(dataset = NULL, as.sparse = TRUE) {
  available_datasets <- c("ica_bone_marrow", "ica_cord_blood")
  if(!is.null(dataset)) {
    if(!(dataset %in% available_datasets)) {
      stop("No dataset found with the specified name, please choose one of the following: \n",
           paste(available_datasets,collapse = "\n"))
    }
  } else {
    message("No dataset name provided, please choose one of the following: \n",
            paste(available_datasets,collapse = "\n"))
    return(invisible(NULL))
  }

  message("Retrieving Human Cell Atlas dataset: ", dataset)
  hub <- ExperimentHub::ExperimentHub()
  base <- "HCAData/"

  ## row and column data
  rdatapath <- paste0(base, dataset, "_rowData.rds")
  rowData <- query(hub, rdatapath)[[1]]

  suppressMessages({
    cdatapath <- paste0(base, dataset, "_colData.rds")
    colData <- query(hub, cdatapath)[[1]]

    ## HDF5, from ExperimentHub:
    h5datapath <- paste0(base, dataset, "_rectangular.h5")
    h5file <- query(hub, h5datapath)[[1]]
  })
  h5array <- HDF5Array(h5file, "counts", as.sparse = as.sparse)

  sce <- SingleCellExperiment(
    list(counts = h5array),
    rowData = rowData,
    colData = colData
  )
  rownames(sce) <- rowData(sce)$ID
  colnames(sce) <- colData(sce)$Barcode
  return(sce)
}
